srml/contract/src/lib.rs

```diff
@@ -121,14 +122,87 @@ pub trait ComputeDispatchFee<Call, Balance> {
 	fn compute_dispatch_fee(call: &Call) -> Balance;
 }
 
-#[derive(Encode,Decode,Clone,Debug)]
 /// Information for managing an acocunt and its sub trie abstraction.
 /// This is the required info to cache for an account
-pub struct AccountInfo {
-	/// unique ID for the subtree encoded as a byte
+#[derive(Encode, Decode)]
+pub enum ContractInfo<T: Trait> {
+	Alive(AliveContractInfo<T>),
+	Tombstone(TombstoneContractInfo<T>),
+}
+
+impl<T: Trait> ContractInfo<T> {
+	/// If contract is alive then return some alive info
+	pub fn get_alive(self) -> Option<AliveContractInfo<T>> {
+		if let ContractInfo::Alive(alive) = self {
+			Some(alive)
+		} else {
+			None
+		}
+	}
+	/// If contract is alive then return some reference to alive info
+	pub fn as_alive(&self) -> Option<&AliveContractInfo<T>> {
+		if let ContractInfo::Alive(ref alive) = self {
+			Some(alive)
+		} else {
+			None
+		}
+	}
+	/// If contract is alive then return some mutable reference to alive info
+	pub fn as_alive_mut(&mut self) -> Option<&mut AliveContractInfo<T>> {
+		if let ContractInfo::Alive(ref mut alive) = self {
+			Some(alive)
+		} else {
+			None
+		}
+	}
+}
+
+#[derive(Encode, Decode)]
+pub struct AliveContractInfo<T: Trait> {
+	/// Unique ID for the subtree encoded as a bytes vector
 	pub trie_id: TrieId,
-	/// the size of stored value in octet
+	/// The size of stored value in octet
 	pub storage_size: u64,
+	/// The code associated with a given account.
+	pub code_hash: CodeHash<T>,
+	pub rent_allowance: BalanceOf<T>,
+	pub deduct_block: T::BlockNumber,
+}
+
+impl<T: Trait> Clone for AliveContractInfo<T> {
+	fn clone(&self) -> Self {
+		AliveContractInfo {
+			trie_id: self.trie_id.clone(),
+			storage_size: self.storage_size.clone(),
+			code_hash: self.code_hash.clone(),
+			rent_allowance: self.rent_allowance.clone(),
+			deduct_block: self.deduct_block.clone(),
+		}
+	}
+}
+
+impl<T: Trait> Eq for AliveContractInfo<T> {}
+
+impl<T: Trait> PartialEq for AliveContractInfo<T> {
+	fn eq(&self, other: &Self) -> bool {
+		self.trie_id == other.trie_id
+			&& self.storage_size == other.storage_size
+			&& self.rent_allowance == other.rent_allowance
+			&& self.code_hash == other.code_hash
+			&& self.deduct_block == other.deduct_block
+	}
+}
```

Comment:
This (and `Clone` as well) implementation a bit worrisome since we can break the `PartialEq` contract by adding a new field.

I understand that `Encode` and `Decode` is really useful, but maybe we should go by replacing `<T: Trait>` with `<CodeHash, Balance, BlockNumber>` and using `derive(Clone, Eq, PartialEq, ..)` for now ?